### PR TITLE
466012: Update to consume openSUSE Tumbleweed and Python 3.8

### PR DESCRIPTION
**Pull request by rorytorneymf**
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=477072

Moving to openSUSE Tumbleweed, which allows us to install Python 3.8+, which we need in order to run this version of the Talon library:

https://github.com/rorytorneymf/talon/releases/tag/1.6.0-with-8138ea9a604f037f47f544dfb805f13d26696275-reverted

### CODE REVIEW

File: src/main/docker/Dockerfile [hunk 1]
```diff
@@ -17,23 +17,25 @@
 # Docker registry
 ARG DOCKER_HUB_PUBLIC=docker.io
 
-FROM ${DOCKER_HUB_PUBLIC}/cafapi/opensuse-jdk11:3.4.6
+FROM ${DOCKER_HUB_PUBLIC}/cafapi/prereleases:opensuse-jdk11-3.7.0-477072-SNAPSHOT
```
Comment (rorytorneymf) [inline src/main/docker/Dockerfile:20]: Update before merging

Comment (dermot-hardy) [reply src/main/docker/Dockerfile:20]: Actually I don't want to move the main branch of the `jdk11` image to Tumbleweed so don't update this before merging.

**Review by dermot-hardy — APPROVED** [review]

**Comment by dermot-hardy** [comment]
We only actually need to go to Python 3.7, right?  Is it possible to adjust this to just go to that version rather than all the way up to 3.10?  I assume 3.7 or 3.8 will be released on Leap relatively soon but it could take quite a while for Leap to get to 3.10.

**Comment by rorytorneymf** [comment]
> We only actually need to go to Python 3.7, right? Is it possible to adjust this to just go to that version rather than all the way up to 3.10? I assume 3.7 or 3.8 will be released on Leap relatively soon but it could take quite a while for Leap to get to 3.10.

I'm not sure, LEAP comes packaged with 3.6, but it doesn't work on that:

https://docs.google.com/spreadsheets/d/15bCWbQSSL0iIp9IAP0_soWhc_QUk6mzFX2wqVFnibas/edit#gid=0 

I will try manually installing 3.7 or 3.8 on a Tumbleweed image to test.

**Comment by rorytorneymf** [comment]
> > We only actually need to go to Python 3.7, right? Is it possible to adjust this to just go to that version rather than all the way up to 3.10? I assume 3.7 or 3.8 will be released on Leap relatively soon but it could take quite a while for Leap to get to 3.10.
> 
> I'm not sure, LEAP comes packaged with 3.6, but it doesn't work on that:
> 
> https://docs.google.com/spreadsheets/d/15bCWbQSSL0iIp9IAP0_soWhc_QUk6mzFX2wqVFnibas/edit#gid=0
> 
> I will try manually installing 3.7 or 3.8 on a Tumbleweed image to test.

@dermot-hardy 

I tried the following:

**Python 3.8 on Tumblweed using a community-provided package**

I tried installing Python 3.8 (https://software.opensuse.org/download/package?package=python3&project=home%3AAlexander_Naumov%3ASSLmigration) on openSUSE Tumbleweed using the following:

```
docker run --entrypoint "/bin/bash" -it cafapi/prereleases:opensuse-jdk11-3.7.0-477072-SNAPSHOT
```

I then ran the following commands individually, in order to select option 2 for this warning:

```
57eff3c4e001:/ # zypper install python3
Loading repository data...
Reading installed packages...
Resolving package dependencies...

Problem: nothing provides 'libnsl.so.2()(64bit)' needed by the to be installed python3-3.7.3-38.6.x86_64
 Solution 1: do not install python3-3.7.3-38.6.x86_64
 Solution 2: break python3-3.7.3-38.6.x86_64 by ignoring some of its dependencies

```

```
zypper -n addrepo https://download.opensuse.org/repositories/home:Alexander_Naumov:SSLmigration/openSUSE_Tumbleweed/home:Alexander_Naumov:SSLmigration.repo
zypper refresh
zypper install python3
```

Can run these as one:

```
zypper -n update && \
zypper -n install python3-devel && \
zypper -n install python3-pip && \
zypper -n install python3-matplotlib && \
zypper -n install zlib-devel && \
zypper -n install python3-numpy-devel && \
zypper -n install python3-lxml && \
zypper -n install python3-scipy && \
zypper -n install gcc-c++ && \
pip install scikit-learn==1.0.1 && \
pip install regex==2022.6.2 && \
pip install -U https://github.com/rorytorneymf/talon/archive/refs/tags/1.6.0-with-8138ea9a604f037f47f544dfb805f13d26696275-reverted.zip  && \
pip install jep==4.0.3 && \
zypper -n clean --all
```

I can then run the python script and confirm that it returns the expected results as needed to get our integration tests to pass.

So, Talon *appears* to work on Python 3.8, however, I will not know for sure until Python 3.8 is officially packaged with openSUSE.

Also, because of the missing dep error running `zypper install python3` above, I cannot run the above commands non-interactively, so I cannot do this in a branch of opensuse-jeptalon-image and test it works end-to-end using Jep.

**Python 3.8 on openSUSE 15.4 using a community-provided package**

https://software.opensuse.org/download/package?package=python3&project=home%3Awicked%3Aqubes-build

```
zypper addrepo https://download.opensuse.org/repositories/home:Ded_Secer:branches:openSUSE:Factory/15.4/home:Ded_Secer:branches:openSUSE:Factory.repo
zypper refresh
zypper install python3
```

Fails when trying to install python3-pip, so didn't  go any further with this:

```
19817892094a:/ # zypper -n install python3-pip
Loading repository data...
Reading installed packages...
Resolving package dependencies...

Problem: the to be installed python3-pip-20.0.2-150400.15.6.noarch requires 'python3-setuptools', but this requirement cannot be provided
  not installable providers: python3-setuptools-44.1.1-150400.1.4.noarch[repo-oss]
 Solution 1: Following actions will be done:
  downgrade of python3-base-3.8.14-lp154.24.1.x86_64 to python3-base-3.6.15-150300.10.30.1.x86_64
  install python3-base-3.6.15-150300.10.30.1.x86_64 from vendor SUSE LLC <https://www.suse.com/>
    replacing python3-base-3.8.14-lp154.24.1.x86_64 from vendor obs://build.opensuse.org/home:Ded_Secer
  downgrade of python3-devel-3.8.14-lp154.24.1.x86_64 to python3-devel-3.6.15-150300.10.30.1.x86_64
  install python3-devel-3.6.15-150300.10.30.1.x86_64 from vendor SUSE LLC <https://www.suse.com/>
    replacing python3-devel-3.8.14-lp154.24.1.x86_64 from vendor obs://build.opensuse.org/home:Ded_Secer
  downgrade of python3-3.8.14-lp154.24.1.x86_64 to python3-3.6.15-150300.10.30.1.x86_64
  install python3-3.6.15-150300.10.30.1.x86_64 from vendor SUSE LLC <https://www.suse.com/>
    replacing python3-3.8.14-lp154.24.1.x86_64 from vendor obs://build.opensuse.org/home:Ded_Secer
  deinstallation of libpython3_8-1_0-3.8.14-lp154.24.1.x86_64
 Solution 2: do not install python3-pip-20.0.2-150400.15.6.noarch
 Solution 3: break python3-pip-20.0.2-150400.15.6.noarch by ignoring some of its dependencies
 ```

**Python 3.7 on Tumbleweed using a community-provided package**

```
zypper addrepo https://download.opensuse.org/repositories/home:jayvdb:python37-static/openSUSE_Tumbleweed/home:jayvdb:python37-static.repo
zypper refresh
zypper install python37
```

Problem with this is there is no python37-* packages available like we have in python38 above, so I had to install the Python 3.10 versions

```
zypper -n refresh && \
    zypper -n update && \
    zypper -n install python3-devel && \
    zypper -n install python3-pip && \
    zypper -n install python3-matplotlib && \
    zypper -n install zlib-devel && \
    zypper -n install python3-numpy-devel && \
    zypper -n install python3-lxml && \
    zypper -n install python3-scipy && \
    zypper -n install gcc-c++ && \
    pip install scikit-learn==1.0.1 && \
    pip install regex==2022.6.2 && \
    pip install -U https://github.com/rorytorneymf/talon/archive/refs/tags/1.6.0-with-8138ea9a604f037f47f544dfb805f13d26696275-reverted.zip  && \
    pip install jep==4.0.3 && \
    zypper -n clean --all
```

I got various errors such as:

```
0c59f26bf678:/ # zypper install python3-pip
Loading repository data...
Reading installed packages...
'python3-pip' not found in package names. Trying capabilities.
Resolving package dependencies...

Problem: the to be installed python310-pip-22.0.4-2.1.noarch requires '/usr/bin/python3.10', but this requirement cannot be provided
  not installable providers: python310-base-3.10.7-2.1.i586[repo-oss]
                   python310-base-3.10.7-2.1.x86_64[repo-oss]
 Solution 1: Following actions will be done:
  deinstallation of python3-3.7.3-38.6.x86_64
  install python310-base-3.10.7-2.1.x86_64 from vendor openSUSE
    replacing python3-base-3.7.3-39.110.x86_64 from vendor obs://build.opensuse.org/home:jayvdb
  deinstallation of python3-devel-3.7.3-39.110.x86_64
 Solution 2: do not ask to install a solvable providing python3-pip
 Solution 3: break python310-pip-22.0.4-2.1.noarch by ignoring some of its dependencies
```

There is also no python-pip or python3-pip package available to install, so I couldn't go any further with this.

**Comment by dermot-hardy** [comment]
I think you've missed the official Tumbleweed Python 3.8 package:
https://software.opensuse.org/package/python38

In fact I see there's an experimental version for Leap 15.4 on that page as well.

To try it you'd need to check in the public key that we should trust, the same way we did for [opensuse-python3-image@v2.0.0](https://github.com/CAFapi/opensuse-python3-image/tree/v2.0.0/src/main/docker) when we were briefly using an experimental package.

**Comment by rorytorneymf** [comment]
> I think you've missed the official Tumbleweed Python 3.8 package: https://software.opensuse.org/package/python38
> 
> In fact I see there's an experimental version for Leap 15.4 on that page as well.
> 
> To try it you'd need to check in the public key that we should trust, the same way we did for [opensuse-python3-image@v2.0.0](https://github.com/CAFapi/opensuse-python3-image/tree/v2.0.0/src/main/docker) when we were briefly using an experimental package.

Thanks, I have updated this to use Python 3.8 now (via `zypper -n install python38`), and all appears to work ok.

**Pull request merged**